wiki/en/en-Command-Line-Options.md

```diff
@@ -52,4 +52,4 @@ You can see all possible options your version supports by starting Jamulus with
 |    `-z` |`--startminimized` | Start minimized | (server only) |
 |       |`--ctrlmidich`     | MIDI controller channel to listen on and control number offset, format: `channel[;offset]` | (client only) see [Tips & Tricks](Tips-Tricks-More#Using-ctrlmidich-for-MIDI-controllers) |
 |       |`--clientname`     | Window title and JACK client name | (client only) |
-|       |`--serverpublicip` | When running your own central server with own slave servers in a private network, specify your public IP with this option to enable clients from the Internet to connect. You will still need proper port forwarding in your router/firewall.| (server only) |
+|       |`--serverpublicip` | Allows servers to be seen on the Internet when listed by a Central Server on a LAN | (server only) [See note](Choosing-a-Server-Type) |
```

Comment:
This is not a requirement when running a Central Server on a LAN.  Rock, Jazz, Classical/Folk, Choir/Barbershop all run on my LAN here (192.168.1.x internal address).  Nothing needed other than a public static IP and port forwarding on the router.

Reply:
OK do we mean, "Allows servers to be seen on the Internet when listed by a Central Server behind a NAT firewall"?

Reply:
Ah, sorry, I've just tested the _explicit_ scenario.  This is when _both_ servers are inside the LAN, behind the NAT wall.

So I've just started up jamulus.drealm.info:21224, port mapped by the router to internal port 55850.  I registered that on jamuluscentral.drealm.info:22824.

This confuses Jamulus Central and completely stops my connecting internally....
![image](https://user-images.githubusercontent.com/1549463/107849047-6dffee80-6df0-11eb-8aa7-6334d35d5613.png)

![image](https://user-images.githubusercontent.com/1549463/107849055-7b1cdd80-6df0-11eb-8eed-28775326918b.png)

I'll try again with the --serverpublicip -- on the registering server.

Reply:
Yes, sorry, this is the whole point of the `--serverpublicip` option as I understand it. Of course, if you are running the directory on your LAN with servers  on that LAN and don't want clients on the WAN in those servers, then it's not needed. 

(Note how the use of the word "Directory" nicely disambiguates "Server"  in this convo - possibly a bit of an edge case example, but still)

Reply:
I'm trying to get a server on my LAN registered with a directory on the LAN _and_ exposed to Jamulus Central.

Reply:
Test scenarios:

* Directory, no "serverpublicip".  Unregistered server, no "serverpublicip", no "centralserver", port not remapped by the router
![image](https://user-images.githubusercontent.com/1549463/107852325-13728c80-6e08-11eb-8afd-5a0563bab7c9.png)
![image](https://user-images.githubusercontent.com/1549463/107852360-3dc44a00-6e08-11eb-8a5b-eeaba20b39fe.png)
Above is as expected
* Directory, no "serverpublicip".  Registered server, no "serverpublicip", "centralserver" as 127.0.0.1:22824, port not remapped by the router
![image](https://user-images.githubusercontent.com/1549463/107852442-9e538700-6e08-11eb-9229-d6eb1a3365b6.png)
![image](https://user-images.githubusercontent.com/1549463/107852471-c347fa00-6e08-11eb-97b4-74f44337562c.png)
It looks like this has a problem - it's showing the address of the registered server as 127.0.0.1 in the directory, which isn't much use...  Worst, the registered server port has been confused with the directory server.  I've no idea how that happens - possibly because the registered server pings Jamulus Explorer, exposing its real public IP address, which is the same as the directory server -- and Jamulus Explorer has nothing else it can do but link by that IP address, despite the port not matching.
```
Feb 13 14:33:27 fs-peter Jamulus-Central8[1079682]: Requested to register entry for 127.0.0.1:56850 (192.168.1.19:56850): Test Server (21124)
```
* Directory, no "serverpublicip".  Registered server, "serverpublicip" set to the 81.174.155.168, "centralserver" as 127.0.0.1:22824, port not remapped by the router
![image](https://user-images.githubusercontent.com/1549463/107852792-94328800-6e0a-11eb-98d8-36f3f47c81c2.png)
![image](https://user-images.githubusercontent.com/1549463/107852814-aa404880-6e0a-11eb-8302-5c60dd47ac15.png)
So as ever, port not remapped by the router _appears_ to work here.
```
Feb 13 14:47:27 fs-peter Jamulus-Central8[1082559]: Requested to register entry for 127.0.0.1:56850 (81.174.155.168:56850): Test Server (21124)
```

What can't happen successfully is to have a second registered server inside my LAN running on port 56850.  Well, internally, it would be okay: the IP:port pair would be unique.  But externally, the serverpublicip IP and port wouldn't, unless the port got mapped.  But, unfortunately, port mapping and directory server ping back don't mix.  So I could turn off the ping back - losing some support for registering external servers - or ensure all registered servers on the LAN, regardless of IP, have a unique port.

Reply:
OK, so the next tests are the same as above but here we'll use port forwarding on the router to see how that affects life.  Not directly related to this but it's still something I'm hoping we'll one day get working with registered servers.

* Directory, no "serverpublicip". Unregistered server, no "serverpublicip", no "centralserver", port remapped by the router
(without forwarding)
![image](https://user-images.githubusercontent.com/1549463/107853490-55063600-6e0e-11eb-9359-e411422ae16a.png)
(with forwarding)
![image](https://user-images.githubusercontent.com/1549463/107853501-694a3300-6e0e-11eb-880c-1771ddb0c93b.png)
And, of course, nothing registered.  Unregistered server log:
```
Feb 13 15:13:47 fs-peter Jamulus-wip[1086790]: - selected port number: 56850
Feb 13 15:13:47 fs-peter Jamulus-wip[1086790]: - maximum number of channels: 4
Feb 13 15:13:47 fs-peter Jamulus-wip[1086790]: - server info: Test Server (21124);West Ealing;224
```
* Directory, no "serverpublicip". Registered server, "serverpublicip" set to the 81.174.155.168, "centralserver" as 127.0.0.1:22824, port remapped by the router
![image](https://user-images.githubusercontent.com/1549463/107853673-39e7f600-6e0f-11eb-998b-1eb501cf1802.png)
![image](https://user-images.githubusercontent.com/1549463/107853705-5e43d280-6e0f-11eb-871b-9913712e302f.png)
The issue we're having is clear from the directory server log
```
Feb 13 15:20:07 fs-peter Jamulus-Central8[1086736]: Requested to register entry for 127.0.0.1:56850 (81.174.155.168:56850): Test Server (21124)
```
* Directory, no "serverpublicip". Registered server, "serverpublicip" set to the 81.174.155.168:21124, "centralserver" as 127.0.0.1:22824, port remapped by the router
Sadly, this fails -- silently, which is bad:
```
Feb 13 15:24:45 fs-peter Jamulus-Central8[1088417]: Requested to register entry for 127.0.0.1:56850 (127.0.0.1:56850): Test Server (21124)
```

Reply:
I also notice that _nothing_ is logged when using `--serverpublicip`, which goes against normal practice for server start up parameters.  I should have spotted that before, when reviewing the code, rather than now, when reviewing the docs....

Reply:
I can't explain the Jamulus Explorer output for the registered server, here, though.

Reply:
Will follow up soon...

Reply:
> I also notice that nothing is logged when using --serverpublicip, which goes against normal practice for server start up parameters. I should have spotted that before, when reviewing the code, rather than now, when reviewing the docs....

I wasn't aware, I'm sorry. I'll fix that. See https://github.com/jamulussoftware/jamulus/pull/993

> Directory, no "serverpublicip". Registered server, "serverpublicip" set to the 81.174.155.168, "centralserver" as 127.0.0.1:22824, port remapped by the router
> Feb 13 15:24:45 fs-peter Jamulus-Central8[1088417]: Requested to register entry for 127.0.0.1:56850 (127.0.0.1:56850): Test Server (21124)

This sounds strange. Can you tell me what versions you are running?
For the new scenario (regular server + directory behind NAT + client from public internet) to work, the following is needed:
- The regular server needs to use Jamulus master code and `--serverpublicip` with the appropriate public IP.
- The directory needs to use Jamulus master (but no special command line options). I suspect that this requirement might not have been met in your test?

This is the scenario which @genesisproject2020 has validated (I also tested behind NAT + Jamulus Explorer as a client).

Reply:
> The directory needs to use Jamulus master

What does this mean?

Reply:
> What does this mean? If you mean the code, then you can see the commit in use in the screen shots.

Sorry, missed that. Everything looks good then.

> The problem only happens when trying to use port forwarding as well. It's just the --serverpublicip syntax doesn't support supplying a public port as well as a public IP.

Ah, right, you mean forwarding a specific port on your public router/firewall to a different internal port. I talked to @genesisproject2020 about adding support for this scenario as well, but this would require a change in the protocol (PROTMESSID_CLM_REGISTER_SERVER). For now, it wasn't deemed necessary.

So yes, the current code/protocol only supports supplying a different IP address. Both internal and external ports have to match.

Reply:
OK so in terms of my suggested wording for the docs (which in this PR is supposed to be a summary, with the [full description here](https://github.com/jamulussoftware/jamuluswebsite/commit/8258ddb4ab912ef3ce12c2894d2807961301c4e6)) - does anything need to change?

Reply:
It needs to say _how to use it_, not what it does.

Reply:
The pattern we have adopted for other things on this page is to summarise the use here and then link to the corresponding docs. If we try to put all the explanation into the table it'll get very cluttered.

Reply:
> `--clientname`     | Window title and JACK client name

So something like `--serverpublicip` | Supply the server public IP address

Reply:
Sorry, how does this work? I've updated that wording but it;s in a separate PR. https://github.com/jamulussoftware/jamuluswebsite/pull/304